build-common/src/org/jetbrains/kotlin/incremental/storage/RelocatableFileToPathConverter.kt

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.incremental.storage
+
+import java.io.File
+
+/**
+ * [FileToPathConverter] which converts a [File] to a Unix-style relative path, which is relative to a base directory ([baseDir]), and vice
+ * versa.
+ *
+ * This is to support build cache relocatability
+ * (https://docs.gradle.org/current/userguide/build_cache_concepts.html#relocatability).
+ */
+class RelocatableFileToPathConverter(private val baseDir: File) : FileToPathConverter {
+
+    private val unixStyleBaseDirPathPrefix = "${baseDir.invariantSeparatorsPath}/"
+
+    override fun toPath(file: File): String {
+        check(file.invariantSeparatorsPath.startsWith(unixStyleBaseDirPathPrefix)) {
+            "The given file '${file.path}' is located outside the base directory '${baseDir.path}'"
+        }
+        return file.relativeTo(baseDir).invariantSeparatorsPath
+    }
+
+    override fun toFile(path: String): File {
+        // Note: The given path is Unix-style but baseDir could be Windows-style; call normalize() so that the style of the returned file's
+        // path is consistent with baseDir.
+        return baseDir.resolve(path).normalize()
```

Comment:
To ensure consistency, should we throw if path is absolute?

Reply:
Yes if `toFile` is an independent API.

However, here `toFile()` is supposed to be called only on the returned value of a `toPath()`, so given `toPath()` is implemented correctly, we will always get a Unix-style relative path here.

If we add this check, we might need to also check that the path is Unix-style, and add similar checks for `baseDir` to and the `file` argument in `toPath()`, so we'll end up having a lot of checks.

Reply:
Kk, leaving it as-is sounds good.